keps/sig-node/3545-improved-multi-numa-alignment/README.md

```diff
@@ -0,0 +1,444 @@
+
+As of now, TopologyManager provides policies to either provide resources from only single numa node or the smallest number of NUMA nodes. TopologyManager is not aware of the NUMA distances and does not take it into consideration during admission decisions. 
+
+This limitation surfaces in multi socket, as well as single socket multi NUMA systems (commonly present in AMD hardware), and might cause a significant performance degradation for latency-critical execution and high-throughput applications, in a case when the Topology Manager schedules a pod on non-adjacent numa nodes.
+
+
+### Goals
+
+* Enable TopologyManager to prefer sets of NUMA nodes with shorter NUMA distances for all TopologyManager policies
+* Preserve all other properties of all policies
+
+## Proposal
+
+### Risks and Mitigations
+
+| Risk                                             | Impact | Mitigation                                                                                                                                                                               |
+|--------------------------------------------------|--------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Bugs in the implementation lead to kubelet crash | High   | Disable the policy option and restart the kubelet. The workload will run but resources allocation can be spread among NUMA nodes which are non-adjacent what can lead to higher latency. |
+
+## Design Details
+
+
+### Proposed Change
+ 
+We propose to
+- add a new flag in Kubelet called `TopologyManagerPolicyOptions` in the kubelet config or command line argument called `topology-manager-policy-options` which allows the user to specify the Topology Manager policy option.
+- add a new topology manager option called `prefer-closest-numa-nodes`; if present, this option will enable further refinements of the existing `restricted` and `best-effort` policies, this option has no effect for `none` and `single-numa-node` policies.
+
+Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
+This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.
+
+To summarize properties of `prefer-closest-numa-nodes` policy:
+
+* Preserve all properties of `best-effort` and `restricted` policies
+* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.
+
+When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.
+Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).
```

Comment:
how do we want to add this dependency? IOW, does cadvisor require this KEP to be accepted before to accept this enhancement, or we can enhance cadvisor in parallel? If this KEP gets delayed for whatever reason, can we have the enhancement in place in cadvisor anyway?

Reply:
My goal was to create PR to `cadvisor` in parallel but don't make it block this proposal if we won't be able to make it in before release.

Reply:
This approach is good enough for me. If we can spell out in the KEP which timelines are looking at to remove the temporary code and depend on cadvisor (e.g.: consider this a beta or a GA blocker) that would be perfect.

Reply:
Graduation criteria was updated with requirement of cadvisor `MachineInfo` update for beta release

Reply:
@fromanirh @knopt I have create a PR to add NUMA distances to `cadvisor` https://github.com/google/cadvisor/pull/3179